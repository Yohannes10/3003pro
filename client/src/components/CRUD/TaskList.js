import React, { useState, useEffect } from "react";
import { Button, Form } from "react-bootstrap";
import "react-datepicker/dist/react-datepicker.css";
import Rating from "react-rating-stars-component"; // Import the Rating component
import Delete from "./Delete";
import "./TaskList.css"

const TaskList = ({
  tasks,
  handleDeleteTask,
  handleEditTask,
  handleToggleTask,
}) => {
  const [ratings, setRatings] = useState({});
  const [departmentalGoals, setDepartmentalGoals] = useState({});

  const handleRatingChange = (taskId, rating) => {
    setRatings((prevRatings) => ({
      ...prevRatings,
      [taskId]: rating,
    }));
  };

  useEffect(() => {
    console.log("Tasks Prop Updated in TaskList:", tasks);
  }, [tasks]);

  console.log("Tasks in TaskList component:", tasks);

  useEffect(() => {
    console.log("Received Tasks in TaskList:", tasks); // Console log for tasks
    // Fetch departmental goals from the server and update the state
    const fetchDepartmentalGoals = async () => {
      try {
        const response = await fetch(
          "http://localhost:8080/departmental-goals"
        );
        if (response.ok) {
          const data = await response.json();
          setDepartmentalGoals(data);
          console.log("Fetched Departmental Goals:", data); // Add this console.log statement
          // Convert the array of departmental goals into an object with goal ID as keys
          const goalsObject = data.reduce((acc, goal) => {
            acc[goal._id] = goal;
            return acc;
          }, {});
          setDepartmentalGoals(goalsObject);
          console.log("Received Tasks in TaskList:", tasks); // Add this console.log statement
          console.log("Fetched Departmental Goals:", goalsObject); // Add this console.log statement
        }
      } catch (error) {
        console.error("Error fetching departmental goals:", error);
      }
    };
    fetchDepartmentalGoals();
  }, []);
  console.log("Received Tasks Prop in TaskList:", tasks);

  if (!tasks || tasks.length === 0) {
    return <h2>Task List is empty</h2>;
  }

  return (
    <div data-testid="task-list-container"> {/* Add data-testid here */}
    <div className="row">
      <div className="row justify-content-center align-items-center">
        {tasks.map((task) => (
          <div className="col-md-4 p-2 my-2" key={task._id}>
            <div className="card rounded-0">
              <div className="card-header d-flex justify-content-between align-items-center">
                <h5>{task.title}</h5>
                <Button
                  onClick={() => handleEditTask(task._id)}
                  className="btn btn-sm"
                >
                  <i className="material-icons"></i>
                  Edit
                </Button>
              </div>
              <div className="card-body">
                {/* Render the departmental goal information */}
              {/*   <p>
                  Departmental Goal:{" "}
                  {departmentalGoals[task.departmentalGoal]?.title}
                </p>
 */}
                <Form.Check
                  type="checkbox"
                  checked={task.completed}
                  onChange={() =>
                    handleToggleTask(task._id, { completed: !task.completed })
                  }
                  style={{
                    border: "2px solid #007bff", // Change border color to blue
                    boxShadow: "0 1px 3px rgba(0, 0, 0, 0.2)", // Add a shadow effect
                    borderRadius: "px", // Add a border radius
                    padding: "4px", // Add some padding for a cleaner look
                    width: "25px",
                  }}
                />
                <p>Description: {task.description}</p>
                <p>Date: {task.date}</p>
                <div>
                  <label>Please Rate Your Performance</label>
                  <Rating
                    count={5}
                    size={24}
                    value={ratings[task._id] || 0}
                    onChange={(newRating) =>
                      handleRatingChange(task._id, newRating)
                    }
                    half={false}
                  />
                </div>
              </div>
              <div className="d-flex justify-content-between align-items-center">
                  <Button
                    onClick={() => handleDeleteTask(task._id)}
                    className="btn btn-sm delete-button" 
                  >
                    Delete
                  </Button>
                </div>
            </div>
          </div>
        ))}
      </div>
    </div>
   </div>
  );
};

export default TaskList;
