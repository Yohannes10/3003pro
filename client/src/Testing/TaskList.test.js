import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect'; // Import the jest-dom matchers
import TaskList from "../components/CRUD/TaskList"

// Define mock tasks for testing
const mockTasks = [
  {
    _id: '1',
    title: 'Task 1',
    description: 'Description 1',
    completed: false,
    date: '2023-08-11',
  },
  // Add more mock tasks as needed
];

// Define mock handler functions
const mockHandleDeleteTask = jest.fn();
const mockHandleEditTask = jest.fn();
const mockHandleToggleTask = jest.fn();

test('renders TaskList component with tasks', () => {
  render(
    <TaskList
      tasks={mockTasks}
      handleDeleteTask={mockHandleDeleteTask}
      handleEditTask={mockHandleEditTask}
      handleToggleTask={mockHandleToggleTask}
    />
  );

  // Check if task titles are rendered
  const taskTitles = mockTasks.map((task) => task.title);
  taskTitles.forEach((title) => {
    const titleElement = screen.getByText(title);
    expect(titleElement).toBeInTheDocument();
  });

  // Check if task descriptions are rendered
  const taskDescriptions = mockTasks.map((task) => task.description);
  taskDescriptions.forEach((description) => {
    const descriptionElement = screen.getByText(/Description 1/);
    expect(descriptionElement).toBeInTheDocument();
  });

  // Check if task checkboxes are rendered
  const taskCheckboxes = mockTasks.map((task) => task.completed);
  taskCheckboxes.forEach((completed) => {
    const checkboxElement = screen.getByRole('checkbox', { checked: completed });
    expect(checkboxElement).toBeInTheDocument();
  });

  // Perform snapshot testing
  expect(screen.getByTestId('task-list-container')).toMatchSnapshot();
});

test('handles checkbox toggle', () => {
  render(
    <TaskList
      tasks={mockTasks}
      handleDeleteTask={mockHandleDeleteTask}
      handleEditTask={mockHandleEditTask}
      handleToggleTask={mockHandleToggleTask}
    />
  );

  // Find the first checkbox and click it
  const firstCheckbox = screen.getAllByRole('checkbox')[0];
  fireEvent.click(firstCheckbox);

  // Verify that the handleToggleTask function was called with the correct taskId and completed status
  expect(mockHandleToggleTask).toHaveBeenCalledWith(mockTasks[0]._id, { completed: true });
});
