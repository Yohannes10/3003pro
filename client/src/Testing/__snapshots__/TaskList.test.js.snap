// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TaskList component with tasks 1`] = `
<div
  data-testid="task-list-container"
>
   
  <div
    class="row"
  >
    <div
      class="row justify-content-center align-items-center"
    >
      <div
        class="col-md-4 p-2 my-2"
      >
        <div
          class="card rounded-0"
        >
          <div
            class="card-header d-flex justify-content-between align-items-center"
          >
            <h5>
              Task 1
            </h5>
            <button
              class="btn btn-sm btn btn-primary"
              type="button"
            >
              <i
                class="material-icons"
              />
              Edit
            </button>
          </div>
          <div
            class="card-body"
          >
            <div
              class=""
              style="border: 2px solid #007bff; box-shadow: 0 1px 3px rgba(0, 0, 0, 0.2); border-radius: px; padding: 4px; width: 25px;"
            >
              <input
                class="form-check-input"
                type="checkbox"
              />
            </div>
            <p>
              Description: 
              Description 1
            </p>
            <p>
              Date: 
              2023-08-11
            </p>
            <div>
              <label>
                Please Rate Your Performance
              </label>
              <div
                class="react-stars-wrapper-01973958556085258"
                style="display: flex;"
              >
                <div
                  aria-label="add rating by typing an integer from 0 to 5 or pressing arrow keys"
                  class="undefined react-stars"
                  style="overflow: hidden; position: relative;"
                  tabindex="0"
                >
                  <span
                    class=""
                    data-forhalf="★"
                    data-index="0"
                    style="position: relative; overflow: hidden; cursor: pointer; display: block; float: left; color: gray; font-size: 24px;"
                  >
                    ★
                  </span>
                  <span
                    class=""
                    data-forhalf="★"
                    data-index="1"
                    style="position: relative; overflow: hidden; cursor: pointer; display: block; float: left; color: gray; font-size: 24px;"
                  >
                    ★
                  </span>
                  <span
                    class=""
                    data-forhalf="★"
                    data-index="2"
                    style="position: relative; overflow: hidden; cursor: pointer; display: block; float: left; color: gray; font-size: 24px;"
                  >
                    ★
                  </span>
                  <span
                    class=""
                    data-forhalf="★"
                    data-index="3"
                    style="position: relative; overflow: hidden; cursor: pointer; display: block; float: left; color: gray; font-size: 24px;"
                  >
                    ★
                  </span>
                  <span
                    class=""
                    data-forhalf="★"
                    data-index="4"
                    style="position: relative; overflow: hidden; cursor: pointer; display: block; float: left; color: gray; font-size: 24px;"
                  >
                    ★
                  </span>
                  <p
                    role="status"
                    style="position: absolute; left: -200rem;"
                  >
                    0
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div
            class="card-footer"
          >
            <button
              class="btn btn-danger btn btn-primary"
              type="button"
            >
              Delete
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
